File: gui/content_factory/batch_master.py
"""
Content Factory — Batch Mastering Lite

Masters multiple songs with ONLY:
  - Dynamics (compression) — even out dynamic range
  - Maximizer (loudness) — consistent LUFS across all songs
  - Imager (stereo width) — consistent stereo image

Does NOT apply EQ because different songs have different tones.
Applying one EQ curve to 20+ songs would destroy tonal balance.

Backend priority:
  1. Rust (PyO3) — longplay.batch_master() with rayon parallel processing
     → Handles unlimited songs, uses all CPU cores
  2. Python fallback — _RealAudioProcessor from chain.py (sequential)
     → Works without compiled Rust backend
"""

from __future__ import annotations

import os
import logging
import threading
from typing import List, Optional, Callable
from pathlib import Path

import numpy as np

from .models import SongEntry, BatchMasterConfig

logger = logging.getLogger(__name__)

# ── Rust backend (preferred — parallel processing via rayon) ──
try:
    from longplay import (
        batch_master as rust_batch_master,
        PyBatchMasterConfig,
        get_parallelism,
    )
    HAS_RUST_BACKEND = True
    logger.info(
        f"[BatchMaster] Rust backend available ({get_parallelism()} threads)"
    )
except ImportError:
    HAS_RUST_BACKEND = False

# Audio I/O
try:
    import soundfile as sf
    HAS_SOUNDFILE = True
except ImportError:
    HAS_SOUNDFILE = False

# Loudness normalization
try:
    import pyloudnorm as pyln
    HAS_PYLOUDNORM = True
except ImportError:
    HAS_PYLOUDNORM = False

# DSP modules — these are CONFIG HOLDERS, not processors
# Audio processing is done by _RealAudioProcessor static methods
try:
    from modules.master.dynamics import Dynamics
    from modules.master.imager import Imager
    from modules.master.maximizer import Maximizer
    from modules.master.limiter import LookAheadLimiterFast
    from modules.master.chain import _RealAudioProcessor
    HAS_DSP = True
except ImportError:
    HAS_DSP = False


class BatchMasterLite:
    """
    Masters a batch of songs using Dynamics + Loudness + Imager only.

    Automatically uses the Rust backend (longplay.batch_master) when available
    for parallel processing across all CPU cores. Falls back to sequential
    Python processing via _RealAudioProcessor from chain.py.

    Usage:
        bm = BatchMasterLite(config)
        bm.master_batch(songs, output_dir, progress_callback)
    """

    def __init__(self, config: Optional[BatchMasterConfig] = None):
        self._config = config or BatchMasterConfig()
        self._dynamics: Optional[Dynamics] = None
        self._imager: Optional[Imager] = None
        self._maximizer: Optional[Maximizer] = None
        self._cancelled = threading.Event()

        if HAS_DSP:
            self._dynamics = Dynamics()
            self._imager = Imager()
            self._maximizer = Maximizer()

    # ─── Public API ──────────────────────────────────────────────

    def cancel(self) -> None:
        """Signal cancellation of current batch."""
        self._cancelled.set()

    @staticmethod
    def has_rust_backend() -> bool:
        """Check if Rust parallel backend is available."""
        return HAS_RUST_BACKEND

    def master_batch(
        self,
        songs: List[SongEntry],
        output_dir: str,
        progress_callback: Optional[Callable[[int, int, str], None]] = None,
    ) -> int:
        """
        Master all songs in batch.

        Uses Rust parallel backend when available, otherwise falls back
        to sequential Python processing.

        Args:
            songs: List of SongEntry to master.
            output_dir: Directory to write mastered files.
            progress_callback: fn(current_index, total, status_msg)

        Returns:
            Number of successfully mastered songs.
        """
        if not self._config.enabled:
            return len(songs)

        os.makedirs(output_dir, exist_ok=True)
        self._cancelled.clear()

        # Try Rust parallel backend first
        if HAS_RUST_BACKEND:
            return self._master_batch_rust(songs, output_dir, progress_callback)

        # Fall back to Python sequential processing
        return self._master_batch_python(songs, output_dir, progress_callback)

    # ─── Rust Backend (parallel via rayon) ────────────────────────

    def _master_batch_rust(
        self,
        songs: List[SongEntry],
        output_dir: str,
        progress_callback: Optional[Callable[[int, int, str], None]] = None,
    ) -> int:
        """Master batch using Rust + rayon parallel processing."""
        total = len(songs)
        threads = get_parallelism()
        logger.info(
            f"[BatchMaster] Rust backend: {total} songs × {threads} threads"
        )

        # Build input/output path lists
        input_paths = []
        output_paths = []
        for song in songs:
            src = song.effective_path
            stem = Path(src).stem
            out = os.path.join(output_dir, f"{stem}_mastered.wav")
            input_paths.append(src)
            output_paths.append(out)

        # Build Rust config
        rust_config = PyBatchMasterConfig(
            dynamics_enabled=self._config.dynamics_enabled,
            imager_enabled=self._config.imager_enabled,
            maximizer_enabled=self._config.maximizer_enabled,
            target_lufs=self._config.target_lufs,
            true_peak_limit=self._config.true_peak_limit,
            dynamics_threshold=self._config.dynamics_threshold,
            dynamics_ratio=self._config.dynamics_ratio,
            imager_width=float(self._config.imager_width),
            maximizer_ceiling=self._config.maximizer_ceiling,
        )

        # Progress callback wrapper
        def rust_progress(completed: int, total_count: int, current_file: str):
            if progress_callback:
                filename = Path(current_file).stem
                progress_callback(
                    completed, total_count, f"Mastering: {filename}"
                )

        # Call Rust batch_master (releases GIL, processes in parallel)
        try:
            results = rust_batch_master(
                input_paths, output_paths, rust_config, rust_progress
            )
        except Exception as e:
            logger.error(f"[BatchMaster] Rust batch failed: {e}, falling back to Python")
            return self._master_batch_python(songs, output_dir, progress_callback)

        # Map results back to songs
        success_count = 0
        for song, result in zip(songs, results):
            if result.success:
                song.mastered_path = result.output_path
                success_count += 1
            else:
                logger.error(
                    f"[BatchMaster] Failed: {song.title} — {result.error}"
                )

        if progress_callback:
            progress_callback(total, total, "Mastering complete")

        logger.info(
            f"[BatchMaster] Rust: {success_count}/{total} mastered "
            f"({threads} threads)"
        )
        return success_count

    # ─── Python Backend (sequential fallback) ─────────────────────

    def _master_batch_python(
        self,
        songs: List[SongEntry],
        output_dir: str,
        progress_callback: Optional[Callable[[int, int, str], None]] = None,
    ) -> int:
        """Master batch using sequential Python DSP (fallback)."""
        success_count = 0
        total = len(songs)

        logger.info(f"[BatchMaster] Python backend: {total} songs (sequential)")

        for i, song in enumerate(songs):
            if self._cancelled.is_set():
                logger.info("[BatchMaster] Cancelled by user")
                break

            if progress_callback:
                progress_callback(i, total, f"Mastering: {song.title}")

            try:
                out_path = self._master_song_python(song, output_dir)
                if out_path:
                    song.mastered_path = out_path
                    success_count += 1
            except Exception as e:
                logger.error(f"Failed to master {song.title}: {e}")

        if progress_callback:
            progress_callback(total, total, "Mastering complete")

        return success_count

    def master_song(self, song: SongEntry, output_dir: str) -> Optional[str]:
        """
        Master a single song (Python backend).

        Returns output file path, or None on failure.
        """
        if not self._config.enabled:
            return song.file_path

        if HAS_RUST_BACKEND:
            return self._master_song_rust(song, output_dir)

        return self._master_song_python(song, output_dir)

    def _master_song_rust(self, song: SongEntry, output_dir: str) -> Optional[str]:
        """Master single song via Rust backend."""
        src = song.effective_path
        stem = Path(src).stem
        out = os.path.join(output_dir, f"{stem}_mastered.wav")

        rust_config = PyBatchMasterConfig(
            dynamics_enabled=self._config.dynamics_enabled,
            imager_enabled=self._config.imager_enabled,
            maximizer_enabled=self._config.maximizer_enabled,
            target_lufs=self._config.target_lufs,
            true_peak_limit=self._config.true_peak_limit,
            dynamics_threshold=self._config.dynamics_threshold,
            dynamics_ratio=self._config.dynamics_ratio,
            imager_width=float(self._config.imager_width),
            maximizer_ceiling=self._config.maximizer_ceiling,
        )

        try:
            results = rust_batch_master([src], [out], rust_config)
            if results and results[0].success:
                return results[0].output_path
            return None
        except Exception as e:
            logger.warning(f"Rust master_song failed: {e}, trying Python")
            return self._master_song_python(song, output_dir)

    def _master_song_python(self, song: SongEntry, output_dir: str) -> Optional[str]:
        """Master a single song using Python DSP."""
        if not HAS_SOUNDFILE:
            logger.warning("soundfile not available — skipping mastering")
            return song.file_path

        # Read audio
        try:
            data, sr = sf.read(song.file_path)
        except Exception as e:
            logger.error(f"Cannot read {song.file_path}: {e}")
            return None

        # Ensure stereo
        if data.ndim == 1:
            data = np.column_stack([data, data])

        logger.info(f"[BatchMaster] Processing: {song.title} "
                     f"({len(data)/sr:.1f}s @ {sr}Hz)")

        # ── Processing pipeline (NO EQ) ──

        # 1. Dynamics (compression)
        if self._config.dynamics_enabled and self._dynamics:
            data = self._apply_dynamics(data, sr)

        # 2. Imager (stereo width)
        if self._config.imager_enabled and self._imager:
            data = self._apply_imager(data, sr)

        # 3. Maximizer (gain push + ceiling)
        if self._config.maximizer_enabled and self._maximizer:
            data = self._apply_maximizer(data, sr)

        # 4. Loudness normalization (LUFS target)
        data = self._normalize_loudness(data, sr)

        # 5. True peak limiting
        data = self._true_peak_limit(data, sr)

        # 6. Final clip
        data = np.clip(data, -1.0, 1.0).astype(np.float32)

        # Write output
        stem = Path(song.file_path).stem
        out_path = os.path.join(output_dir, f"{stem}_mastered.wav")

        try:
            sf.write(out_path, data, sr, subtype='PCM_24')
            logger.info(f"[BatchMaster] Written: {out_path}")
            return out_path
        except Exception as e:
            logger.error(f"Cannot write {out_path}: {e}")
            return None

    # ─── DSP Steps (using _RealAudioProcessor from chain.py) ─────

    def _apply_dynamics(self, data: np.ndarray, sr: int) -> np.ndarray:
        """Apply gentle compression via _RealAudioProcessor.process_dynamics()."""
        if not self._dynamics or not HAS_DSP:
            return data

        try:
            # Configure Dynamics config object (single-band for batch)
            self._dynamics.enabled = True
            self._dynamics.multiband = False
            sb = self._dynamics.single_band
            sb.threshold = self._config.dynamics_threshold  # -18 dB
            sb.ratio = self._config.dynamics_ratio          # 2.5
            sb.attack = 10.0                                # ms
            sb.release = 100.0                              # ms
            sb.knee = 6.0                                   # dB (soft knee)

            # Use _RealAudioProcessor static method (NOT .process())
            return _RealAudioProcessor.process_dynamics(
                data, sr, self._dynamics, intensity=1.0
            )
        except Exception as e:
            logger.warning(f"Dynamics processing failed: {e}")
            return data

    def _apply_imager(self, data: np.ndarray, sr: int) -> np.ndarray:
        """Apply stereo width via _RealAudioProcessor.process_imager()."""
        if not self._imager or not HAS_DSP or data.ndim < 2:
            return data

        try:
            # Configure Imager config object
            # Imager.width uses 0-200 integer scale (100 = original)
            self._imager.enabled = True
            self._imager.width = self._config.imager_width  # 0-200 int

            # Use _RealAudioProcessor static method
            return _RealAudioProcessor.process_imager(
                data, sr, self._imager, intensity=1.0
            )
        except Exception as e:
            logger.warning(f"Imager processing failed: {e}")
            return data

    def _apply_maximizer(self, data: np.ndarray, sr: int) -> np.ndarray:
        """Apply gain maximization via _RealAudioProcessor.process_maximizer()."""
        if not self._maximizer or not HAS_DSP:
            return data

        try:
            # Configure Maximizer config object
            self._maximizer.enabled = True
            self._maximizer.ceiling = self._config.maximizer_ceiling  # -0.3 dB

            # Use _RealAudioProcessor static method
            return _RealAudioProcessor.process_maximizer(
                data, sr, self._maximizer, intensity=1.0
            )
        except Exception as e:
            logger.warning(f"Maximizer processing failed: {e}")
            return data

    def _normalize_loudness(self, data: np.ndarray, sr: int) -> np.ndarray:
        """Normalize to target LUFS using pyloudnorm."""
        if not HAS_PYLOUDNORM:
            return data

        try:
            meter = pyln.Meter(sr)
            current_lufs = meter.integrated_loudness(data)

            if np.isinf(current_lufs) or np.isnan(current_lufs):
                return data

            target = self._config.target_lufs
            return pyln.normalize.loudness(data, current_lufs, target)
        except Exception as e:
            logger.warning(f"Loudness normalization failed: {e}")
            return data

    def _true_peak_limit(self, data: np.ndarray, sr: int) -> np.ndarray:
        """Apply true peak brickwall limiting."""
        if not HAS_DSP:
            # Fallback: simple hard clip at ceiling
            ceiling_linear = 10 ** (self._config.true_peak_limit / 20.0)
            peak = np.max(np.abs(data))
            if peak > ceiling_linear:
                data = data * (ceiling_linear / peak)
            return data

        try:
            limiter = LookAheadLimiterFast(
                ceiling_db=self._config.true_peak_limit,
                lookahead_ms=1.5,
                release_ms=50.0,
            )
            return limiter.process(data, sr)
        except Exception as e:
            logger.warning(f"True peak limiting failed: {e}")
            ceiling_linear = 10 ** (self._config.true_peak_limit / 20.0)
            peak = np.max(np.abs(data))
            if peak > ceiling_linear:
                data = data * (ceiling_linear / peak)
            return data

File: rust/crates/longplay-chain/src/batch_master.rs
//! BatchMasterLite — Parallel batch mastering for Content Factory
//!
//! Masters N songs in parallel using rayon, applying ONLY:
//!   - Dynamics (compression) — even out dynamic range
//!   - Imager (stereo width) — consistent stereo image
//!   - Maximizer (loudness push + ceiling)
//!   - Loudness normalization (RMS-based LUFS approximation)
//!   - True peak hard clip
//!
//! Does NOT apply EQ — different songs have different tonal balance.
//! Applying one EQ curve to many songs would destroy their character.

use longplay_core::AudioBuffer;
use longplay_core::conversions::db_to_linear_f;
use longplay_dsp::dynamics::Dynamics;
use longplay_dsp::imager::Imager;
use longplay_dsp::maximizer::Maximizer;

use rayon::prelude::*;
use std::sync::atomic::{AtomicBool, AtomicUsize, Ordering};
use std::sync::Arc;

/// Configuration for batch mastering
#[derive(Debug, Clone)]
pub struct BatchMasterConfig {
    pub dynamics_enabled: bool,
    pub imager_enabled: bool,
    pub maximizer_enabled: bool,
    pub target_lufs: f32,
    pub true_peak_limit: f32,
    pub dynamics_threshold: f64,
    pub dynamics_ratio: f64,
    pub imager_width: f32,
    pub maximizer_ceiling: f32,
}

impl Default for BatchMasterConfig {
    fn default() -> Self {
        Self {
            dynamics_enabled: true,
            imager_enabled: true,
            maximizer_enabled: true,
            target_lufs: -14.0,
            true_peak_limit: -1.0,
            dynamics_threshold: -18.0,
            dynamics_ratio: 2.5,
            imager_width: 100.0,   // 100 = original width (0-200 scale)
            maximizer_ceiling: -0.3,
        }
    }
}

/// Result of mastering a single song
#[derive(Debug, Clone)]
pub struct BatchMasterResult {
    pub input_path: String,
    pub output_path: String,
    pub success: bool,
    pub error: Option<String>,
    pub duration_sec: f64,
}

/// Progress callback type for batch mastering
/// (completed_count, total_count, current_file)
pub type BatchProgressCallback = Box<dyn Fn(usize, usize, &str) + Send + Sync>;

/// Master a single audio buffer through the lite pipeline (no EQ).
///
/// Signal flow:
///   1. Dynamics (compression)
///   2. Imager (stereo width)
///   3. Maximizer (gain push + ceiling)
///   4. Loudness normalization (RMS → target LUFS)
///   5. True peak hard clip
fn process_buffer_lite(
    buffer: &AudioBuffer,
    sample_rate: u32,
    config: &BatchMasterConfig,
) -> AudioBuffer {
    if buffer.is_empty() || buffer[0].is_empty() {
        return buffer.clone();
    }

    let mut buf = buffer.clone();
    let sr = sample_rate as i32;

    // Stage 1: Dynamics (compression)
    if config.dynamics_enabled {
        let mut dynamics = Dynamics::new();
        dynamics.set_bypass(false);
        dynamics.set_multiband(false);
        dynamics.single_band_mut().set_threshold(config.dynamics_threshold);
        dynamics.single_band_mut().set_ratio(config.dynamics_ratio);
        dynamics.single_band_mut().set_attack(10.0);   // ms
        dynamics.single_band_mut().set_release(100.0);  // ms
        buf = dynamics.process(&buf, sr);
    }

    // Stage 2: Imager (stereo width)
    if config.imager_enabled && buf.len() >= 2 {
        let mut imager = Imager::new();
        imager.set_bypass(false);
        imager.set_width(config.imager_width);
        imager.process(&mut buf, sr);
    }

    // Stage 3: Maximizer (gain push + ceiling)
    if config.maximizer_enabled {
        let mut maximizer = Maximizer::new();
        maximizer.set_bypass(false);
        maximizer.set_ceiling(config.maximizer_ceiling);
        maximizer.process(&mut buf, sr);
    }

    // Stage 4: Loudness normalization (RMS-based LUFS approximation)
    let loudness_gain = estimate_rms_gain(&buf, config.target_lufs);
    for channel in &mut buf {
        for sample in channel.iter_mut() {
            *sample *= loudness_gain;
        }
    }

    // Stage 5: True peak hard clip
    let ceiling_linear = db_to_linear_f(config.true_peak_limit);
    for channel in &mut buf {
        for sample in channel.iter_mut() {
            if *sample > ceiling_linear {
                *sample = ceiling_linear;
            } else if *sample < -ceiling_linear {
                *sample = -ceiling_linear;
            }
        }
    }

    buf
}

/// RMS-based gain estimation to approximate target LUFS.
fn estimate_rms_gain(buffer: &AudioBuffer, target_lufs: f32) -> f32 {
    if buffer.is_empty() || buffer[0].is_empty() {
        return 1.0;
    }

    let mut sum_sq = 0.0f64;
    let mut total_samples = 0usize;

    for channel in buffer {
        for &sample in channel {
            sum_sq += (sample as f64) * (sample as f64);
        }
        total_samples += channel.len();
    }

    if total_samples == 0 {
        return 1.0;
    }

    let rms = (sum_sq / total_samples as f64).sqrt();
    if rms <= 1e-10 {
        return 1.0;
    }

    let rms_db = 20.0 * rms.log10();
    let gain_db = (target_lufs as f64) - rms_db;
    let gain_db = gain_db.clamp(-20.0, 20.0);

    10.0_f64.powf(gain_db / 20.0) as f32
}

/// Master a single file: read → process → write.
fn master_single_file(
    input_path: &str,
    output_path: &str,
    config: &BatchMasterConfig,
) -> BatchMasterResult {
    // Read input (supports WAV, MP3, FLAC, OGG via longplay_io)
    let (buffer, info) = match longplay_io::read_audio_any(input_path) {
        Ok(result) => result,
        Err(e) => {
            return BatchMasterResult {
                input_path: input_path.to_string(),
                output_path: output_path.to_string(),
                success: false,
                error: Some(format!("Read error: {}", e)),
                duration_sec: 0.0,
            };
        }
    };

    // Ensure stereo
    let buffer = if buffer.len() == 1 {
        longplay_io::mono_to_stereo(&buffer)
    } else {
        buffer
    };

    // Process through lite pipeline
    let processed = process_buffer_lite(&buffer, info.sample_rate as u32, config);

    // Write output as 24-bit WAV
    let bit_depth = if info.bit_depth > 0 && info.bit_depth <= 32 {
        info.bit_depth
    } else {
        24
    };

    match longplay_io::write_audio(output_path, &processed, info.sample_rate, bit_depth) {
        Ok(_) => BatchMasterResult {
            input_path: input_path.to_string(),
            output_path: output_path.to_string(),
            success: true,
            error: None,
            duration_sec: info.duration_seconds,
        },
        Err(e) => BatchMasterResult {
            input_path: input_path.to_string(),
            output_path: output_path.to_string(),
            success: false,
            error: Some(format!("Write error: {}", e)),
            duration_sec: info.duration_seconds,
        },
    }
}

/// Batch master multiple songs in parallel using rayon.
///
/// # Arguments
/// * `input_paths` — List of input audio file paths
/// * `output_paths` — Corresponding output file paths (must be same length)
/// * `config` — Batch mastering configuration
/// * `cancelled` — Atomic flag to signal cancellation
/// * `progress_callback` — Optional progress callback (completed, total, current_file)
///
/// # Returns
/// Vector of results, one per input file.
pub fn batch_master(
    input_paths: &[String],
    output_paths: &[String],
    config: &BatchMasterConfig,
    cancelled: Option<Arc<AtomicBool>>,
    progress_callback: Option<&BatchProgressCallback>,
) -> Vec<BatchMasterResult> {
    assert_eq!(
        input_paths.len(),
        output_paths.len(),
        "Input and output path counts must match"
    );

    let total = input_paths.len();
    let completed = Arc::new(AtomicUsize::new(0));
    let cancel_flag = cancelled.unwrap_or_else(|| Arc::new(AtomicBool::new(false)));

    // Pair input/output paths for parallel iteration
    let pairs: Vec<(&String, &String)> = input_paths.iter().zip(output_paths.iter()).collect();

    let results: Vec<BatchMasterResult> = pairs
        .par_iter()
        .map(|(input, output)| {
            // Check cancellation
            if cancel_flag.load(Ordering::Relaxed) {
                return BatchMasterResult {
                    input_path: input.to_string(),
                    output_path: output.to_string(),
                    success: false,
                    error: Some("Cancelled".to_string()),
                    duration_sec: 0.0,
                };
            }

            let result = master_single_file(input, output, config);

            // Update progress
            let done = completed.fetch_add(1, Ordering::Relaxed) + 1;
            if let Some(cb) = progress_callback {
                cb(done, total, input);
            }

            result
        })
        .collect();

    results
}

/// Get the number of available CPU threads for parallel processing
pub fn available_parallelism() -> usize {
    std::thread::available_parallelism()
        .map(|n| n.get())
        .unwrap_or(4)
}
